src/nnet3/nnet-chain-training.cc

```diff
@@ -167,6 +188,14 @@ void NnetChainTrainer::UpdateParamsWithMaxChange() {
                   << "UpdatableComponent; change this code.";
       BaseFloat max_param_change_per_comp = uc->MaxChange();
       KALDI_ASSERT(max_param_change_per_comp >= 0.0);
+      if (nnet_config.backstitch_training_scale > 0.0 &&
```

Comment:
You need to do a much better job, via comments, of explaining what is going on here.   Even I don't understand this.  It's possible that the code structure here is the problem, and it will have to be changed so that it's easier to understand.  

Here is my proposal.

We take UpdateParamsWithMaxChange out of the class (probably declare it in nnet-utils.h), remove the is_backstitch_step  argument, take out the ScaleNnet() from it [that can be done separately by the 
calling code], and give it this interface:
```
/**
       This function does the operation '*nnet += scale * delta_nnet', while respecting
       any max-parameter-change (max-param-change) specified in the updatable
       components, plus the global max-param-change specified as 'max_param_change'.

       With max-changes taken into account, the operation of this function is equivalent
       to the following, although it's done more efficiently:
        \code
             Nnet temp_nnet(delta_nnet);
             ScaleNnet(1.0 / max_change_scale, &temp_nnet);
             [ Scale down parameters for each component of temp_nnet as needed so
               their Euclidean norms do not exceed their per-component max-changes ]
             [ Scale down temp_nnet as needed so its Euclidean norm does not exceed
               the global max-change ]
             ScaleNnet(max_change_scale, &temp_nnet);  // undo the previous scaling.
             AddNnet(temp_nnet, scale, nnet);
        \endcode

       @param [in] delta_nnet  The copy of '*nnet' neural network that contains 
                                  the proposed change in parameters.  Normally this will previously
                                  have been set to: (delta_nnet = 
                                  parameter-derivative-on-current-minibatch * learning-rate per parameter),
                                  with any natural gradient applied as specified in the components; but
                                  this may be different if momentum or backstitch are used.
      @param [in] max_param_change  The global max-param-change specified on the 
                                  command line (e.g. 2.0), which specifies the largest change allowed
                                  to '*nnet' in Euclidean norm.  If <= 0, no global max-param-change will
                                  be enforced, but any max-change values specified in the components
                                  will still be enforced;  see UpdatableComponent::MaxChange(), and 
                                  search for 'max-change' in the configs or nnet3-info output).
      @param [in] max_change_scale  This value, which will normally be 1.0, is used to scale
                                  all per-component max-change values and the global 'max_param_change',
                                 before applying them (so we use 'max_param_change * uc->MaxChange()'
                                 as the per-component max-change, and 'max_param_change * max_change_scale'
                                 as the global max-change).
      @param [in] scale  This value, which will normally be 1.0, is a scaling factor used when adding 
                                 to 'nnet', applied after any max-changes.
       @param [in,out] nnet  The nnet which we add to.
*/
void UpdateNnetWithMaxChange(const Nnet &delta_nnet, 
                                              BaseFloat max_param_change,
                                              BaseFloat max_change_scale,
                                              BaseFloat scale, Nnet *nnet);
```

You may have to have the backstitch code in a separate code branch from the regular/momentum
training code, to keep the simple case simple.